### ISSUE TITLE
Default ipl.dol not loaded if alternate DOL cannot be loaded.

### ISSUE THREAD

**Issue by yo1dog** [redolution/iplboot#16]
If my SD card has only `ipl.dol`, and I hold any button on startup (perhaps accidentally), the onboard IPL is booted. I would expect `ipl.dol` to be booted as it is the "default".

I believe this describes the most common use-case: I just want to always boot into Swiss and don't care about the shortcuts.

We could update the logic such that it first attempts to load the selected shortcut, falling back to `ipl.dol`. However, since we search over multiple devices, there are two implementation options. Assuming A button is held:

1. For each device, attempt to load `a.dol`. Then, for each device, attempt to load `ipl.dol`. (iterate over devices twice)
2. For each device, attempt to load `a.dol` and then `ipl.dol`. (iterate over devices once)

Current device order is: sdb, sda, sd2. If sdb contains `ipl.dol` and sda contains `a.dol`, no.1 would load `sda://a.dol`, while no.2 would load `sdb://ipl.dol`.

I can see arguments for both. However, there is only a functionality distinction if the user has both A) multiple devices connected and B) DOLs across those devices. I assume this is very unlikely for the vast majority of users. I suggest we go with no.2 as it is easier to implement and more performant (currently we mount and unmout each device).

**Comment by 9ary** [comment]
Yes that's a problem with the current logic. That said, it does conveniently allow you to boot the original IPL by holding down a button that doesn't have anything associated with it. Falling back to ipl.dol would break that.

**Comment by yo1dog** [comment]
True. Though I’d assume that would also be a rarely used feature. Two solutions come to mind:

1. Create DOL that just exits/restarts immediately which would cause onboard IPL to boot. Can assign it to a shortcut.
2. Have a dedicated skip button. Could return early skipping basically all of iplboot (device mounting, looking for files, etc.)

I like 2 as it solves a more general class of problem. Could use D-left. Or some button combination like X & Y.

**Comment by 9ary** [comment]
I also think 2 would be best. A special DOL would just make things more complicated for no reason.

**Comment by yo1dog** [comment]
Agreed. Any preference on how to activate skip (button combo vs dpad)? If we use dpad then I’d say let’s drop support for using the other directions for shortcuts. They are kind of weird anyway.

**Comment by 9ary** [comment]
Agreed on the dpad. The start button and the reset button on the console itself are also good candidates. I think both might be a good idea.

**Comment by yo1dog** [comment]
I vote d-pad left. Start is a good shortcut button.

Didn't even know you could use the reset button. Could reset interfere with downstream software like Swiss or GBI? For example, if I were to not release reset until after Swiss loaded.

Are there any established "skip" button combinations that we could follow?

I'll open two PRs. One to adopt `ipl.dol` fallback. Another to add skip functionality.

**Comment by yo1dog** [comment]
Aside: Just realized this issue is necessary to resolve as it blocks hold-button-on-boot features of downstream software. For example, Swiss allows holding L trigger on boot to force ignore component cable and holding B to disable DVD.

**Comment by 9ary** [comment]
> Didn't even know you could use the reset button. Could reset interfere with downstream software like Swiss or GBI? For example, if I were to not release reset until after Swiss loaded.

I'm pretty sure most homebrew just ignores the reset button. It should be safe with the original IPL as well.

> Aside: Just realized this issue is necessary to resolve as it blocks hold-button-on-boot features of downstream software. For example, Swiss allows holding L trigger on boot to force ignore component cable and holding B to disable DVD.

The stock IPL also allows you to hold A to force it into the menu when a disc is inserted.

**Comment by Extrems** [comment]
GBI uses the reset button to control power to the Game Boy Player.

**Comment by yo1dog** [comment]
Derp. Just realized that we only really care about how onboard IPL reacts to reset button given that skipping always reboots to onboard IPL.

**Comment by 9ary** [comment]
By the way, you should join #gc-forever on efnet. We all hang out there.

**Comment by yo1dog** [comment]
Can't join the efnet servers. Are they down right now?

**Comment by 9ary** [comment]
They're working just fine.
